src/metatrain/soap_bpnn/model.py

```diff
@@ -311,6 +326,72 @@ def forward(
         soap_features = self.layernorm(soap_features)
         features = self.bpnn(soap_features)
 
+        if self.long_range:
+            # slightly painful because:
+            # - the features are split per center type
+            # - we have to recompute the edge vectors again outside of featomic
+
+            # first, send center_type to the samples dimension and make sure the
+            # ordering is the same as in the systems
+            merged_features = (
+                metatensor.torch.sort(
+                    features.keys_to_samples("center_type"), axes="samples"
+                )
+                .block()
+                .values
+            )
+
+            # calculate the distances between all atoms
+            # TODO: replace by something smarter once `register_autograd_neighbors`
+            # is fixed
+            edge_vectors = torch.concatenate(
+                [
+                    system.positions[
+                        system.get_neighbor_list(self.nl_options).samples.values[:, 1]
+                    ]
+                    - system.positions[
+                        system.get_neighbor_list(self.nl_options).samples.values[:, 0]
+                    ]
+                    + system.get_neighbor_list(self.nl_options)
+                    .samples.values[:, 2:]
+                    .to(system.cell.dtype)
+                    @ system.cell
+                    for system in systems
+                ]
+            )
+            distances = torch.sqrt(torch.sum(edge_vectors**2, dim=-1))
+
+            long_range_features_tensor = self.long_range_featurizer(
+                systems, merged_features, distances
+            )
```

Comment:
Shouldn't this only be done for the systems with no pbc? If we do this for all system we loose the NlogN scaling and it becomes ultra slow.